src/unix/udp.c

```diff
@@ -616,23 +639,19 @@ static int uv__setsockopt_maybe_char(uv_udp_t* handle,
   if (val < 0 || val > 255)
     return -EINVAL;
 
-  if (handle->flags & UV_HANDLE_IPV6)
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IPV6,
-                   option6,
-                   &arg,
-                   sizeof(arg));
-  else
-    r = setsockopt(handle->io_watcher.fd,
-                   IPPROTO_IP,
-                   option4,
-                   &arg,
-                   sizeof(arg));
-
-  if (r)
-    return -errno;
+/*
+ * On Solaris and derivatives such as SmartOS, the length of socket options
+ * is always sizeof(int), except for the IP_MULTICAST_LOOP and IP_MULTICAST_TTL
+ * options.
+ */
+#if defined(__sun)
```

Comment:
I like that you separated the ipv4 and ipv6 functions, but I still think it's better to handle special cases in their respective functions rather than doing so in the generic one.

Thanks!